modules/api/pkg/handler/v2/provider/gcp.go

```diff
@@ -106,6 +164,278 @@ func DecodeGCPSubnetworksNoCredentialReq(c context.Context, r *http.Request) (in
+	}
+	return credentials.ServiceAccount, nil
+}
+
+func ListProjectGCPDiskTypes(presetProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter, withProject bool) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		var (
+			req       GCPCommonReq
+			zone      string
+			projectID string
+			sa        string
+		)
+
+		if !withProject {
+			vmReq, ok := request.(GCPVMReq)
+			if !ok {
+				return nil, utilerrors.NewBadRequest("invalid request")
+			}
+			zone = vmReq.Zone
+			req = vmReq.Request
+		} else {
+			projectReq, ok := request.(GCPProjectVMReq)
+			if !ok {
+				return nil, utilerrors.NewBadRequest("invalid request")
+			}
+			zone = projectReq.Zone
+			projectID = projectReq.ProjectID
+			req = projectReq.Request.Request
+		}
+
+		if err := req.Validate(); err != nil {
+			return nil, utilerrors.NewBadRequest(err.Error())
+		}
+
+		sa = req.ServiceAccount
+		var err error
+		if len(req.Credential) > 0 {
+			sa, err = getSAFromPreset(ctx, userInfoGetter, presetProvider, req.Credential, projectID)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return providercommon.ListGCPDiskTypes(ctx, sa, zone)
+	}
+}
+
+func ListProjectGCPZones(presetProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter, seedGetter provider.SeedsGetter, withProject bool) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		var (
+			req        GCPCommonReq
+			projectID  string
+			datacenter string
+			sa         string
+		)
+
+		if !withProject {
+			dcReq, ok := request.(GCPDatacenterReq)
+			if !ok {
+				return nil, utilerrors.NewBadRequest("invalid request")
+			}
+			req = dcReq.Request
+			datacenter = dcReq.DC
+		} else {
+			projectReq, ok := request.(GCPProjectDatacenterReq)
+			if !ok {
+				return nil, utilerrors.NewBadRequest("invalid request")
+			}
+
+			projectID = projectReq.ProjectRequest.ProjectID
+			datacenter = projectReq.DC
+			req = projectReq.ProjectRequest.Request
+		}
+
+		if err := req.Validate(); err != nil {
+			return nil, utilerrors.NewBadRequest(err.Error())
+		}
+
+		sa = req.ServiceAccount
+		var err error
+		if len(req.Credential) > 0 {
+			sa, err = getSAFromPreset(ctx, userInfoGetter, presetProvider, req.Credential, projectID)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		userInfo, err := userInfoGetter(ctx, projectID)
+		if err != nil {
+			return nil, err
+		}
+
+		return providercommon.ListGCPZones(ctx, userInfo, sa, datacenter, seedGetter)
+	}
+}
+
+func ListProjectGCPNetworks(presetProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter, withProject bool) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		var (
+			req       GCPCommonReq
+			projectID string
+			sa        string
+		)
+
+		if !withProject {
+			listReq, ok := request.(GCPProjectCommonReq)
```

Comment:
This should probably be `GCPCommonReq` instead of `GCPProjectCommonReq` since you're in the `!withProject` part of the if condition. The `else` branch also needs to be switched over to be project-aware.